feat: Ctrl+Tab cycling to keep within essentials/regular tabs only

junnjiee opened the pull request:
Note: This PR is a continuation from #11199

## Overview
This PR implements Ctrl+Tab cycling within Essential or Workspace tabs, depending on which tab is currently shown.

preference variable is named `cycleByAttribute` (in `browser.yaml`)

If both `sortByRecentlyUsed` and `cycleByAttribute` are both toggled on, using Ctrl+Tab will default to cycling by recently used. But, using Ctrl+PgUp/PgDn will still activate this feature, as it accesses the `advanceSelectedTab` function directly.

## Implementation Details

### `tabbox.js` (patch file)
- uses `"zen-essential"` attribute to filter tabs.

### `main.inc.xhtml` (patch file)
- checkbox to toggle the feature under Settings > General > Tabs
- warning message below the checkbox element, hidden by default

### `main.js` (patch file)
- `toggleCtrlTabFeatureWarning()` -> shows the warning message when both `sortByRecentlySelected` and `cycleByAttribute` are both toggled on
<img width="579" height="161" alt="image" src="https://github.com/user-attachments/assets/2be7fce5-8b44-49cf-b884-25f276876eb1" />

## Testing
- `browser_tabs_cycle_by_attribute.js` -> tests that tab cycles within essential/regular tabs when `cycleByAttribute` is true. This is achieved by checking if the tab cycles back to itself (example: 3 tabs, advance 3 times, so starting tab should cycle back to itself)
---
Issues/Discussions requesting for the feature: #4787, #6316, #9119.

junnjiee committed:
feat: ctrl+tab cycles within essential/workspace tabs only

junnjiee committed:
add test for ctrl+tab cycling by attribute

junnjiee commented:
@mr-cheffy Thanks for reviewing my code in #11199! This PR includes your requested changes
- localization is moved to `zen-preferences.ftl`
- used lazy pref getter in `tabbox.js` using `XPCOMUtils.defineLazyPreferenceGetter()`

junnjiee committed:
Merge branch 'dev' into ctrltab-scoped-cycling

mr-cheffy commented on src/browser/components/preferences/main-inc-xhtml.patch:
```diff
@@ -1,8 +1,21 @@
+     <checkbox id="ctrlTabRecentlyUsedOrder" data-l10n-id="ctrl-tab-recently-used-order"
+               preference="browser.ctrlTab.sortByRecentlyUsed"/>
+ 
++    <checkbox id="ctrlTabCycleByAttribute" data-l10n-id="ctrl-tab-cycle-by-attribute"
```
I'd move this into zen's "tab management" settings section, and avoid making patches

mr-cheffy commented on src/browser/components/preferences/main-js.patch:
```diff
@@ -1,8 +1,16 @@
+   { id: "browser.warnOnQuitShortcut", type: "bool" },
+   { id: "browser.tabs.warnOnOpen", type: "bool" },
+   { id: "browser.ctrlTab.sortByRecentlyUsed", type: "bool" },
++  { id: "browser.ctrlTab.cycleByAttribute", type: "bool" },
```
This can be done in zen-settings.js:

https://github.com/zen-browser/desktop/blob/5bd0a01d5369d3bc5cc3f2481319b997c0ca703e/src/browser/components/preferences/zen-settings.js#L1057-L1138

mr-cheffy commented on src/browser/components/preferences/main-js.patch:
```diff
@@ -23,3 +31,40 @@ index 46283d3eacfea0f419afa3bdfd2a51e411502a20..4fe229706eb0145a21042b8b3cca3236
+       Services.prefs.clearUserPref("browser.ctrlTab.migrated");
+     });
++
++    Preferences.get("browser.ctrlTab.sortByRecentlyUsed").on(
```
Same as the comment above

mr-cheffy commented on src/toolkit/content/widgets/tabbox-js.patch:
```diff
@@ -50,3 +68,35 @@ index cfe2da6e199667bd668f117cc8972212c7f57da2..bdd7dcc26139202e6e31afde47dc4d87
++          newTab = this.allTabs.find(tabFilter);
+         } else {
+-          newTab = this.allTabs.findLast(tab => tab.visible);
++          newTab = this.allTabs.findLast(tabFilter);
```
We shouldn't really apply this filter here, when the tab is hidden we want to get out of it, even if its outside of our "rules"

mr-cheffy commented on prefs/firefox/browser.yaml:
```diff
@@ -83,3 +83,6 @@
 - name: browser.tabs.closeWindowWithLastTab
   value: false
+
+- name: browser.ctrlTab.cycleByAttribute
```
Please name this zen.tabs.* and move it to the correct zen's config file where the other zen.tabs.* prefs are

mr-cheffy reviewed (changes requested):
Looks good! Some small changes

junnjiee committed:
move settings UI related code to /src/zen, revert patches for main.inc.xhtml and main.js, use simple tab filter when current tab is hidden

junnjiee commented:
Included your requested changes with the most recent commit! @mr-cheffy 

## Key Changes
- reverted patches for `main.inc.xhtml` and `main.js`, so they shouldn't show up in files changed anymore
- UI related logic is now implemented in `zenTabsManagement.inc.html` and `zen-settings.js`
- when tab is hidden, the tab filter used is `tab => tab.visible`

### Updated UI
Toggle now resides in Tab Management > Workspaces
<img width="877" height="401" alt="image" src="https://github.com/user-attachments/assets/3df30d82-fb1a-42d4-afb8-3312efb6d687" />

mr-cheffy committed:
chore: Cleanup and add extra pref, b=no-bug, c=tests, tabs

mr-cheffy reviewed (approved):
LGTM, one thing though: Would it be possible for users to find it annoying if all of they sudden they can't ctrl+Tab outside of the essentials area? e.g. if they enabled it without thinking.

Shouldn't we move the user out of the essentials area, **always**? But that's for a different PR, I guess

junnjiee commented:
Good point! Do you think this behavior should an option, or a default, unchangeable one?

I personally would like cycling within essentials since the commands Alt+1 to 8 only covers up to 8 essential tabs.
